src/OrchardCore.Modules/OrchardCore.Autoroute/Handlers/AutorouteContentHandler.cs

```diff
@@ -0,0 +1,39 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Routing;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Handlers;
+using OrchardCore.ContentManagement.Routing;
+
+namespace OrchardCore.Autoroute.Handlers
+{
+    public class AutorouteContentHandler : ContentHandlerBase
+    {
+        private readonly IAutorouteEntries _autorouteEntries;
+
+        public AutorouteContentHandler(IAutorouteEntries autorouteEntries)
+        {
+            _autorouteEntries = autorouteEntries;
+        }
+
+        public override Task GetContentItemAspectAsync(ContentItemAspectContext context)
+        {
+            return context.ForAsync<ContentItemMetadata>(metadata =>
+            {
+                // When a content item is contained we provide different route values when generating urls.
+                if (_autorouteEntries.TryGetEntryByContentItemId(context.ContentItem.ContentItemId, out var entry) &&
+                    !string.IsNullOrEmpty(entry.ContainedContentItemId))
+                {
+                    metadata.DisplayRouteValues = new RouteValueDictionary {
```

Comment:
one side effect is that if contained item routing is not enabled, generating a route for it with a url helper will generate a standard route (i.e. /Contents/Item/id) which will 404